tests/linearizability/linearizability_test.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linearizability
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/anishathalye/porcupine"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// minimalQPS is used to validate if enough traffic is send to make tests accurate.
+	minimalQPS = 100.0
+	// maximalQPS limits number of requests send to etcd to avoid linearizability analysis taking too long.
+	maximalQPS = 200.0
+	// failpointTriggersCount
+	failpointTriggersCount = 60
+	// waitBetweenFailpointTriggers
+	waitBetweenFailpointTriggers = time.Second
+)
+
+func TestLinearizability(t *testing.T) {
+	testRunner.BeforeTest(t)
+	tcs := []struct {
+		name      string
+		failpoint Failpoint
+		config    e2e.EtcdProcessClusterConfig
+	}{
+		{
+			name:      "KillClusterOfSize1",
+			failpoint: KillFailpoint,
+			config: e2e.EtcdProcessClusterConfig{
+				ClusterSize: 1,
+			},
+		},
+		{
+			name:      "KillClusterOfSize3",
+			failpoint: KillFailpoint,
+			config: e2e.EtcdProcessClusterConfig{
+				ClusterSize: 3,
+			},
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			failpoint := FailpointConfig{
+				failpoint:           tc.failpoint,
+				count:               failpointTriggersCount,
+				waitBetweenTriggers: waitBetweenFailpointTriggers,
+			}
+			traffic := trafficConfig{
+				minimalQPS:  minimalQPS,
+				maximalQPS:  maximalQPS,
+				clientCount: 8,
```

Comment:
minor comment: please consider to make this configurable, such as adding a field `clientCount` into the struct.

Reply:
Struct `tcs` is not set in stone, we can always add it when we decide to parameterize tests based on it.